src/Rules/SyntaxErrorInQueryAssemblerRule.php

```diff
@@ -0,0 +1,181 @@
+<?php
+
+declare(strict_types=1);
+
+namespace staabm\PHPStanDba\Rules;
+
+use PhpParser\Node;
+use PhpParser\Node\Expr\CallLike;
+use PhpParser\Node\Expr\MethodCall;
+use PhpParser\Node\Expr\New_;
+use PhpParser\Node\Name\FullyQualified;
+use PHPStan\Analyser\Scope;
+use PHPStan\Rules\Rule;
+use PHPStan\Rules\RuleError;
+use PHPStan\Rules\RuleErrorBuilder;
+use PHPStan\ShouldNotHappenException;
+use PHPStan\Type\Constant\ConstantArrayType;
+use PHPStan\Type\Constant\ConstantStringType;
+use PHPStan\Type\ObjectType;
+use PHPStan\Type\VerbosityLevel;
+use staabm\PHPStanDba\QueryReflection\QueryReflection;
+
+/**
+ * @implements Rule<CallLike>
+ *
+ * @see SyntaxErrorInQueryAssemblerRuleTest
+ */
+final class SyntaxErrorInQueryAssemblerRule implements Rule
+{
+    /**
+     * @var array<array{string, string, list<int>}>
+     */
+    private $classMethods;
+
+    /**
+     * @var QueryReflection
+     */
+    private $queryReflection;
+
+    /**
+     * @param list<string> $classMethods
+     */
+    public function __construct(array $classMethods)
+    {
+        $this->classMethods = [];
+        foreach ($classMethods as $classMethod) {
+            sscanf($classMethod, '%[^::]::%[^#]#%[0-9,]', $className, $methodName, $arrayArgPositions);
+            if (! \is_string($className) || ! \is_string($methodName)) {
+                throw new ShouldNotHappenException('Invalid classMethod definition');
+            }
+            if ($arrayArgPositions !== null) {
+                $arrayArgPositions = array_map('intval', explode(',', strval($arrayArgPositions)));
+            } else {
+                $arrayArgPositions = [];
+            }
+            $this->classMethods[] = [$className, $methodName, $arrayArgPositions];
+        }
+        $this->queryReflection = new QueryReflection();
+    }
```

Comment:
I am not sure its worth to have this method this kind of configurable.

do we have identical api contracts on classes outside of doctrine/dbal?
I am wondering whether we should instead have 3 rules which map the identical needs of the 3 doctrine-dbal use-cases?

Reply:
If we were to implement them as separate rules with no arguments, I think they would be sufficiently similar that it would make sense to have an abstract base class with child classes that just define which arguments are column name/value arrays. Let me know if you'd like me to make that change, and if you have any suggestions for class names.

I don't know of any interfaces like this outside of Doctrine/DBAL. I have my own wrappers, but they follow the same format, e.g.
* truncate
* delete#1
* replace#1
* insert#1
* update#1,2

I can't imagine any other use cases. Maybe more automated SELECT queries, but probably better to use a fully-fledged query builder at that point.

Reply:
I think it could make sense to have a single rule which can be used for any api which takes a table name as a string.
this could be a use case worth supporting for a separate rule (could be a separate PR though).

for the doctrine dba delete,insert,update api I think we should have a rule dedicated for that api, without configuration but hardcoded signatures

Reply:
Sure, I can make some changes. Just to confirm, instead of:
```neon
    -   
        class: staabm\PHPStanDba\Rules\SyntaxErrorInQueryAssemblerRule
        tags: [phpstan.rules.rule]
        arguments:
            classMethods:
                - 'Doctrine\DBAL\Connection::insert#1'
                - 'Doctrine\DBAL\Connection::delete#1'
                - 'Doctrine\DBAL\Connection::update#1,2'
                - 'My\Connection::truncate'
                - 'My\Connection::insert#1'
                - 'My\Connection::replace#1'
                - 'My\Connection::delete#1'
                - 'My\Connection::update#1,2'
```
You want:
```neon
    -   
        class: staabm\PHPStanDba\Rules\DoctrineErrorInInsertMethodRule
        tags: [phpstan.rules.rule]
    -   
        class: staabm\PHPStanDba\Rules\DoctrineErrorInDeleteMethodRule
        tags: [phpstan.rules.rule]
    -   
        class: staabm\PHPStanDba\Rules\DoctrineErrorInUpdateMethodRule
        tags: [phpstan.rules.rule]
    -   
        class: staabm\PHPStanDba\Rules\ErrorInQueryAssemblerMethodRule
        tags: [phpstan.rules.rule]
        arguments:
            classMethods:
                - 'My\Connection::truncate'
                - 'My\Connection::insert#1'
                - 'My\Connection::replace#1'
                - 'My\Connection::delete#1'
                - 'My\Connection::update#1,2'
```
with the arguments hard-coded for the Doctrine classes? Or do you want Doctrine users to be able to cover all three methods with a single class, e.g. `DoctrineErrorInQueryAssemblerMethodRule`?

Reply:
I am fine with a single doctrine rule, or 3 separate ones.

If we have the doctrine rule we don't need the SyntaxErrorInQueryAssemblerRule anymore, right?

Reply:
At least for my use case, we'd still need the more generic rule, since I have my own wrappers (as enumerated by the `My\Connection` examples). But if you don't want that in the upstream repo, I can remove it.

Reply:
Ok, got it now. Your wrappers class wrapps the doctrine api itself, or just uses the same style?

Reply:
See https://github.com/staabm/phpstan-dba/pull/529/files#r1112353477

Reply:
Sorry for the slow response. I can usually only get to this stuff on the weekends.

It looks like you were able to finish this up in a way that you were happy with. I've confirmed that it works for me. Thanks so much for your help and patience!